enhancements/rhcos/extensions.md

```diff
@@ -0,0 +1,96 @@
+---
+title: support-for-coreos-extensions
+authors:
+  - "@cgwalters"
+reviewers:
+  - "@ashcrow"
+approvers:
+  - "@ashcrow"
+  - "@crawford"
+  - "@imcleod"
+  - "@runcom"
+creation-date: 2020-04-21
+last-updated: 2020-04-21
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Support For CoreOS extensions
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement proposes a MachineConfig fragment like:
+```
+extensions:
+  - usbguard
+```
+
+This is the OpenShift version of the [Fedora CoreOS extension system tracker](https://github.com/coreos/fedora-coreos-tracker/issues/401).
+
+That will add additional software onto the host - but this software will still be versioned with the host (included as part of the OpenShift release payload) and upgraded with the cluster.
+
+## Motivation
+
+We have requests to ship things like `usbguard` that are needed to meet compliance in some scenarios (and `usbguard` is very useful on bare metal), but `usbguard` also makes no sense to ship in an OS for the public cloud.
+
+The `rpm-ostree` project has had as its goal from the start to re-cast RPMs as "operating system extensions" - much like how e.g. Firefox and its extensions work.  By default it operates as a pure image system, but packages can be layered on (and overridden) client side.
+
+OpenShift is already making use of this today for the [realtime kernel](https://github.com/openshift/enhancements/blob/master/enhancements/support-for-realtime-kernel.md).
+
+We propose continuing the trail blazed with the RT kernel by adding additional RPMs to `machine-os-content` that aren't part of the base OS "image", but can be installed later.  This is how `kernel-rt` works today; we added a `kernel-rt.rpm` that is committed to the container.  In the future though, we may instead ship a separate `machine-os-extensions` container, or something else.  The only API stable piece here is the `MachineConfig` field (same as for the `kernelType: realtime`).
```

Comment:
> This is how `kernel-rt` works today; we added a `kernel-rt.rpm` that is committed to the container.

How does this work with base OS versioning? I guess in the existing case you must be doing this in the `machine-os-content` build (or something similarly-low-level) so the build process knows what RHEL/kernel version it is targeting? But what about something like the CNI binary problem? To be able to use this feature to install CNI binaries, we would need to build separate RHEL7 and RHEL8 RPMs and then be able to cause the correct one to be installed on a per-node basis.

Actually, I think the proposal as written is ambiguous about whether you're proposing "there will be a single container in the release image (`machine-os-content`?) that contains RPMs that can be installed as extensions" or "it will be possible to install RPMs that come from any container in the release image". I guess you probably meant the former since there's no discussion of how the RPMs would be found... But that makes it not as good for solving the CNI / https://github.com/coreos/fedora-coreos-tracker/issues/354 problem.

(Although, I guess, we could install the RPMs into "higher-level" images like `origin-multus-route-override-cni`, and then `machine-os-content` or some other "lower-level" image could extract the correct ones from those images and copy them into itself, like it does to get `hyperkube` from the origin image...)

Anyway, I feel like https://github.com/coreos/fedora-coreos-tracker/issues/354 should be explicitly either a Goal or a Non-Goal...

Reply:
I think there's really two levels to this proposal, and I guess I need to cleanly separate them.

The first part is really about the customer UX for adding things that are *not* conceptually on by default, such as `usbguard`.

But what this does quickly bleed into is using such a mechanism for default parts of OpenShift, such as `openvswitch`.

But were we to ship ovs that way, I think it wouldn't be something an administrator could choose in the MCO `extensions` field - it'd be an implementation detail of the clusterNetwork choice.

And the same would be true for the CNI binaries.

Does that makes sense?

Reply:
We don't want to expose CNI plugins / system-vs-containerized OVS as end-user options, but I don't think it makes sense to have the MCO try to figure out what networking RPMs to install by analyzing the network config itself. It seems like it would make more sense to have CNO tell MCO what RPMs it needs, somehow. That may or may not make sense as part of the same mechanism as for the end-user-selectable extensions.

Reply:
OK yep, we're in sync on that then.